consensus_inputs.go

```diff
@@ -0,0 +1,258 @@
+package f3
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"slices"
+	"time"
+
+	"github.com/filecoin-project/go-f3/certs"
+	"github.com/filecoin-project/go-f3/certstore"
+	"github.com/filecoin-project/go-f3/ec"
+	"github.com/filecoin-project/go-f3/gpbft"
+	"github.com/filecoin-project/go-f3/internal/clock"
+	"github.com/filecoin-project/go-f3/manifest"
+	lru "github.com/hashicorp/golang-lru/v2"
+	"github.com/ipfs/go-cid"
+	"go.opentelemetry.io/otel/metric"
+)
+
+type gpbftInputs struct {
+	manifest  *manifest.Manifest
+	certStore *certstore.Store
+	ec        ec.Backend
+	verifier  gpbft.Verifier
+	clock     clock.Clock
+
+	ptCache *lru.Cache[string, cid.Cid]
+}
+
+func newInputs(manifest *manifest.Manifest, certStore *certstore.Store, ec ec.Backend,
+	verifier gpbft.Verifier, clk clock.Clock) gpbftInputs {
+	cache, err := lru.New[string, cid.Cid](256) // keep a bit more than 2x max ECChain size
+	if err != nil {
+		// panic as it only depends on the size
+		panic(fmt.Errorf("could not create cache: %w", err))
+	}
+
+	return gpbftInputs{
+		manifest:  manifest,
+		certStore: certStore,
+		ec:        ec,
+		verifier:  verifier,
+		clock:     clk,
+		ptCache:   cache,
+	}
+}
+
+func (h *gpbftInputs) getPowerTableCIDForTipset(ctx context.Context, tsk gpbft.TipSetKey) (cid.Cid, error) {
+	sTSK := string(tsk)
+	ptCid, ok := h.ptCache.Get(sTSK)
+	if ok {
+		return ptCid, nil
+	}
+
+	pt, err := h.ec.GetPowerTable(ctx, tsk)
+	if err != nil {
+		return cid.Undef, fmt.Errorf("getting power table to compute CID: %w", err)
+	}
+	ptCid, err = certs.MakePowerTableCID(pt)
+	if err != nil {
+		return cid.Undef, fmt.Errorf("computing power table CID: %w", err)
+	}
+
+	h.ptCache.Add(sTSK, ptCid)
+	return ptCid, nil
+}
+
+func (h *gpbftInputs) collectChain(ctx context.Context, base ec.TipSet, head ec.TipSet) ([]ec.TipSet, error) {
```

Comment:
This function would collect the entire chain that _could_ be finalised irrespective of max chain length of 100. Right?

Specifically, during bootstrap this function would, at best, collect a chain of 900, then a chain of 800, 700, and so on even though each instance would only use at most 100 tipsets of the returned chain as proposal.

Is that the most efficient we can be here? How fast/expensive is `GetParent`?

Reply:
It isn't that expensive in comparison to getting the power table, it could be more efficient but it is tricky given null rounds.